{
 "nbformat": 4,
 "nbformat_minor": 0,
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3"
  },
  "language_info": {
   "name": "python"
  }
 },
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# Spatial transcriptomics exercise\n",
    "\n",
    "Please hand in the completed ipynb and a html or pdf version of the notebook by email to daria.romanovskaia@helmholtz-munich.de by **27.01.2026**.\n",
    "\n",
    "The exercise was kindly provided by louis.kummerle@helmholtz-munich.de\n",
    "\n",
    "For any qustions please contact the tutors of the seminar (Daria Romanovskaia, Benedikt Roth, Sarah Ouologuem, Xiatong Fu) on Slack.\n",
    "\n",
    "You will get familiar to spatial transcriptomics analyses by looking at data from the human heart.\n",
    "\n",
    "There is a total of 50 points that can be achieved.\n",
    "\n",
    "## Installing and running\n",
    "The exercise can be implemented in this google colab notebook. If you want to work on colab, please create a copy of the notebook. Alternatively you can download the ipynb notebook and work on your local jupyter server. For the local setup to work, please install python 3.10 and jupyter on your laptop.\n",
    "\n",
    "In both cases you need to install the squidpy package."
   ],
   "metadata": {
    "id": "i3vfVjoCzBey"
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "qj7rs7i1xq_y"
   },
   "outputs": [],
   "source": [
    "!pip install squidpy"
   ]
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Import packages"
   ],
   "metadata": {
    "id": "KKwlgW5wzaOx"
   }
  },
  {
   "cell_type": "code",
   "source": [
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import squidpy as sq\n",
    "import pandas as pd\n",
    "import zipfile\n",
    "\n",
    "## for downloading\n",
    "from pathlib import Path\n",
    "from urllib.request import urlretrieve"
   ],
   "metadata": {
    "id": "E80Y6kYYy1iu"
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Exercise 1: deconvolution of spot based data from the human heart\n",
    "\n",
    "The data for this exercise comes from the version 2 of the human heart cell atlas published in this paper:\n",
    "\n",
    "*Spatially resolved multiomics of human cardiac niches*\n",
    "* https://www.nature.com/articles/s41586-023-06311-1\n",
    "* https://www.heartcellatlas.org/\n",
    "\n",
    "\n",
    "In the paper, the authors characterized the cells of the cardiac conduction system. The data was obtained using the spot based Visium platform on the one hand and using single nucleus RNA-seq.\n",
    "\n",
    "**Aim of this exercise: understand and reproduce figure 1 c-f of the paper**\n",
    "\n",
    "![image on the nature website](https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41586-023-06311-1/MediaObjects/41586_2023_6311_Fig1_HTML.png?as=webp)\n",
    "\n",
    "Download the exercise data"
   ],
   "metadata": {
    "id": "HtyqpvYc0kT2"
   }
  },
  {
   "cell_type": "code",
   "source": [
    "DATA_DIR = Path(\"./data\")\n",
    "DATA_DIR.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "urls = [\n",
    "    # Ex 1.2\n",
    "    \"https://cellgeni.cog.sanger.ac.uk/heartcellatlas/v2/SAN_aCMs_lognormalised.h5ad\",\n",
    "    \"https://cellgeni.cog.sanger.ac.uk/heartcellatlas/v2/AVN_aCMs_lognormalised.h5ad\",\n",
    "    # Ex 1.5\n",
    "    \"https://cellgeni.cog.sanger.ac.uk/heartcellatlas/visium/OCT/HCAHeartST10659160.h5ad\",\n",
    "    \"https://cellgeni.cog.sanger.ac.uk/heartcellatlas/visium/OCT/HCAHeartST13228103.h5ad\",\n",
    "]\n",
    "\n",
    "for url in urls:\n",
    "    print(f\"Downloading {url}...\")\n",
    "    urlretrieve(url, DATA_DIR / Path(url).name)"
   ],
   "metadata": {
    "id": "T51t9wPCy_Ww"
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Exercise 1.1: Describe the UMAPs in the paper (5 points)\n",
    "\n",
    "Please read the [paper](https://www.nature.com/articles/s41586-023-06311-1) and describe what is shown in Figure 1c) and 1d)?\n",
    "\n",
    "a) what kind of data is the UMAP built from?\n",
    "\n",
    "b) which cells were selected for visualization?\n",
    "\n",
    "c) cells of which cell type are highlighted in panels c-d?\n",
    "\n"
   ],
   "metadata": {
    "id": "hUiAoyBI1nAS"
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "_Please insert your solution here_"
   ],
   "metadata": {
    "id": "BWs0xDTR0iOZ"
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Exercise 1.2: Reproduce the UMAPs from the paper (2 points)\n",
    "Reproduce the plot in Figure 1c) and 1d) using data in `AVN_aCMs_lognormalised.h5ad` and `SAN_aCMs_lognormalised.h5ad`. The celltype annotation is contained in the `obs` data frame of these anndata objects."
   ],
   "metadata": {
    "id": "_zSs0TEeEuns"
   }
  },
  {
   "cell_type": "code",
   "source": [
    "## Solution\n",
    "\n"
   ],
   "metadata": {
    "id": "mrrOTXy7Cepo"
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Exercise 1.3: cell type deconvolution using cell2location (10 points)\n",
    "Please read the cell2location [paper](https://www.nature.com/articles/s41587-021-01139-4) and the cell2location [documentation](https://cell2location.readthedocs.io/en/latest/index.html).\n",
    "\n",
    "a) what is the input to cell2location?\n",
    "\n",
    "b) what are the key steps in cell2location and which functions in the code do they correspond to?\n",
    "\n",
    "c) what is the output of cell2location?"
   ],
   "metadata": {
    "id": "9iTm2Qs1h7yl"
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "_Please insert your solution here_"
   ],
   "metadata": {
    "id": "XDLrsnxr0tGO"
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Exercise 1.4: cell type signature for spatial deconvolution (12 points)\n",
    "\n",
    "Now we turn to the spot based spatial transcriptomics data that was generated using the Visum technology. This data was analysed using cell2location (see introductory talk).\n",
    "\n",
    "We followed the code of the publication to learn the cell type signatures of the snRNA-seq data. The code is available here:\n",
    "https://github.com/Teichlab/HCA_Heart_ver2/blob/main/Visium/1_cell2location/reference-signature_revision_SAN-AVN.ipynb\n",
    "\n",
    "The following code chunk downloads and extracts the data.\n"
   ],
   "metadata": {
    "id": "VWXQBdz9IV_t"
   }
  },
  {
   "cell_type": "code",
   "source": [
    "url = \"https://hmgubox2.helmholtz-muenchen.de/index.php/s/XPN7fy5GQY3aZ6z/download\"\n",
    "urlretrieve(url, DATA_DIR / 'cell2location.zip')\n",
    "with zipfile.ZipFile(DATA_DIR / \"cell2location.zip\", 'r') as zip_ref:\n",
    "    zip_ref.extractall(DATA_DIR)"
   ],
   "metadata": {
    "id": "T7H-5US_zGrs"
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "The output files `sc.h5ad`, `model.pt`, `inf_aver.csv` are available here:\n",
    "`DATA_DIR / \"cell2location\" / \"reference-signatures\" / <\"AVN\"/\"SAN\">`. When you work on google colab, you can see the files when you click on the files folder icon on the left."
   ],
   "metadata": {
    "id": "PizAC3u_d7vu"
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "**Our goal now is to verify the cell type signatures:**\n",
    "\n",
    "a) Find the top 10 DE genes for the follwing 3 groups in AVN: P cells, bundle cells, and other cells.\n",
    "\n",
    "b) Generate a dotplot of the marker genes over the 3 groups.\n",
    "\n",
    "c) Based on the dotplot select 10 marker genes that are specific for 1 (or 2) of the 3 groups.\n",
    "\n",
    "d) Create a heatmap or a table to visualize and check the `means_per_cluster_mu_fg` of the different states in AVN and discuss if the learned signatures fit to the marker expressions."
   ],
   "metadata": {
    "id": "HIAYO_WvidSr"
   }
  },
  {
   "cell_type": "code",
   "source": [
    "# solution a-b"
   ],
   "metadata": {
    "id": "mZetlAiHjQtu"
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "code",
   "source": [
    "# solution c"
   ],
   "metadata": {
    "id": "8Tg-DGPdjrK4"
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "code",
   "source": [
    "# solution d"
   ],
   "metadata": {
    "id": "M_Wg620BjvKn"
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "\n",
    "### Exercise 1.5: Reproduce figure 1f) (6 points)\n",
    "Using these celltype signatures, cell2location has been applied to the Visium data sets in `HCAHeartST10659160.h5ad` for the first (left) panel of figure 1f) and `HCAHeartST13228103.h5ad` for the second and third panel of figure 1f). In each file, the `obsm` data frame contains the results of the cell2location deconvolution. Please refer to the [documentation](https://cell2location.readthedocs.io/en/latest/cell2location.html) of the `export_posterior` function to identify the relevant column that carries the information on the mean abundance per cell type.\n",
    "\n",
    "Copy the abundance estimates of the cell types shown in the panels of Figure 1f) as new columns of the anndata `obs` data frame.\n",
    "\n",
    "Identify the scanpy plotting function that is best suited for visualization of spatial data from the [documentation](https://scanpy.readthedocs.io/en/latest/api/plotting.html).\n",
    "\n",
    "a) Use this function to plot the spatial distribution of the cell type abundance estimates created by cell2location in the three leftmost panels of figure 1f).\n",
    "\n",
    "b) Based on your three plots and the [paper](https://www.nature.com/articles/s41586-023-06311-1): explain which of the cell types of the cardiac conduction system are localized in which physiological regions?"
   ],
   "metadata": {
    "id": "xVv85-_uI1ez"
   }
  },
  {
   "cell_type": "code",
   "source": [
    "# solution"
   ],
   "metadata": {
    "id": "d04-hNEHMMwG"
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Exercise 2: spatial statistics on cell type resolved data\n",
    "\n",
    "In this part of the exercise you will look at cell type resolved data from human heart tissue that was profiled on the 10X genomics Xenium platform."
   ],
   "metadata": {
    "id": "W-i5eYjWyHvu"
   }
  },
  {
   "cell_type": "code",
   "source": [
    "url_dict = {\n",
    "    # Ex 2\n",
    "    \"cell_feature_matrix.h5\": \"https://hmgubox2.helmholtz-muenchen.de/index.php/s/6mPbYEb53xR3mKy/download\",\n",
    "    \"cells_annotated.csv\": \"https://hmgubox2.helmholtz-muenchen.de/index.php/s/NiaBtWAEjrDDN8g/download\",\n",
    "}\n",
    "\n",
    "for file in url_dict.keys():\n",
    "    print(f\"Downloading {file}...\")\n",
    "    urlretrieve(url_dict[file], DATA_DIR / file)"
   ],
   "metadata": {
    "id": "QzgHn1KOnQOW"
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "The next code chunk loads the data and plots the cell type annotation on the spatial coordinates. The cell types are annotated as:\n",
    "\n",
    "| Annotation | Cell type |\n",
    "|------------|-----------|\n",
    "| CM | cardiomyocytes |\n",
    "| EC | endothelial cells |\n",
    "| FB | fibroblasts |\n",
    "| Leuko | leukocytes |\n",
    "| NC | neuronal cells |\n",
    "| PC | pericytes |\n",
    "| SMC | smooth muscle cells |"
   ],
   "metadata": {
    "id": "wPZluVcgoqGP"
   }
  },
  {
   "cell_type": "code",
   "source": [
    "adata = sc.read_10x_h5(filename = \"%s/cell_feature_matrix.h5\" % DATA_DIR)\n",
    "cells = pd.read_csv(\"%s/cells_annotated.csv\" % DATA_DIR)\n",
    "adata = adata[cells.index]\n",
    "adata.obs = cells.copy()\n",
    "adata.obsm[\"spatial\"] = adata.obs[[\"y_centroid\", \"x_centroid\"]].copy().to_numpy()\n",
    "\n",
    "sq.pl.spatial_scatter(\n",
    "    adata, color=\"annotation\", shape=None, figsize=(16, 5)\n",
    ")"
   ],
   "metadata": {
    "id": "DIUjKYLMm34e"
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Exercise 2.1: Identify neighborhood enrichment (5 points)\n",
    "In this exercise the aim is to identfy whether specific cell types are preferentially localized next to cells of specific other cell types.\n",
    "\n",
    "a) Check out the squidpy [Xenium tutorial](https://squidpy.readthedocs.io/en/stable/notebooks/tutorials/tutorial_xenium.html) to run the neighborhood enrichment analysis using the anndata object that we just loaded and `cluster_key=\"annotation\"`. Create a heatmap plot and describe the result:\n",
    " - what does the color code indicate?\n",
    " - what is the interpretation of positive and negative values?\n",
    " - which is the highest value that you observe and how do you interpret it?\n",
    "\n",
    "b) Plot the spatial distribution of the three cell types that are most frequent in vessel walls. What do you observe? How is this related to the results of part a)?\n"
   ],
   "metadata": {
    "id": "3MqdNXiVe9Ke"
   }
  },
  {
   "cell_type": "code",
   "source": [
    "# solution a"
   ],
   "metadata": {
    "id": "cxSoICf6p9k7"
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "code",
   "source": [
    "# solution b"
   ],
   "metadata": {
    "id": "Q0ma9flgqEAj"
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Exercise 2.2: co-occurrence analysis (5 points)\n",
    "A more general analysis of co-occurence patterns can be performed when not only the direct neighbors but also cells at greater distances are considered.\n",
    "\n",
    "a) Check out the squidpy [Xenium tutorial](https://squidpy.readthedocs.io/en/stable/notebooks/tutorials/tutorial_xenium.html) to run the co-occurrence analysis using the anndata object that we just loaded and `cluster_key=\"annotation\"`. Focus this analysis on smooth muscle cells (SMC).\n",
    "\n",
    "b) Create the co-occurence plot for SMCs. Describe the plot: what is shown on each of the axes? How do you interprete the result for SMCs? Which tissue structure could explain this observation?\n",
    "\n",
    "_Note_: you can speed up the computations by restricting the analyses to a smaller patch on the image, for example by subsetting the ann data object with this expression `(adata.obs[\"x_centroid\"] > 4000) & (adata.obs[\"x_centroid\"] < 6000) & (adata.obs[\"y_centroid\"] > 5000) & (adata.obs[\"y_centroid\"] < 10000)`"
   ],
   "metadata": {
    "id": "x7FJKX12WpMp"
   }
  },
  {
   "cell_type": "code",
   "source": [
    "# solution a"
   ],
   "metadata": {
    "id": "QpDYe20UXWAq"
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "code",
   "source": [
    "# solution b"
   ],
   "metadata": {
    "id": "EHC6CLJSYY5N"
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Exercise 2.3: Identify genes with spatial gene expression (5 points)\n",
    "\n",
    "Use the squidpy function [`squidpy.gr.autocorr`](https://squidpy.readthedocs.io/en/stable/api/squidpy.gr.spatial_autocorr.html) to identify three different types  spatially patterned genes:\n",
    "\n",
    "a) Find the top 3 genes with the highest Moran's I\n",
    "\n",
    "b) Find the top 3 genes with the highest Moran's I after filtering for genes that are expressed in at least 10% of cells\n",
    "\n",
    "c) Find the top 3 genes with the lowest Moran's I after filtering for genes that are expressed in at least 10% of cells\n",
    "\n",
    "Use the squidpy function [`squidpy.pl.spatial_scatter`](https://squidpy.readthedocs.io/en/stable/api/squidpy.pl.spatial_scatter.html) to plot the spatial expression pattern of each of the top 3 genes from a)-c).\n"
   ],
   "metadata": {
    "id": "ThsNK7O7fHOC"
   }
  },
  {
   "cell_type": "code",
   "source": [
    "# solution"
   ],
   "metadata": {
    "id": "voOKl7aEfd0m"
   },
   "execution_count": null,
   "outputs": []
  }
 ]
}
